google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/MessageDispatcher.java

```diff
@@ -276,7 +275,13 @@ public void processReceivedMessages(List<com.google.pubsub.v1.ReceivedMessage> r
     logger.log(
         Level.FINER, "Received {0} messages at {1}", new Object[] {responseMessages.size(), now});
 
+    try {
+      flowController.reserve(responseMessages.size(), totalByteCount);
+    } catch (FlowController.FlowControlException unexpectedException) {
+      throw new IllegalStateException("Flow control unexpected exception", unexpectedException);
+    }
     messagesWaiter.incrementPendingMessages(responseMessages.size());
+
     Iterator<AckHandler> acksIterator = ackHandlers.iterator();
```

Comment:
I know this is a request to change code outside of the scope of your fix, but your fix brought to my attention that it's a little hard to follow this method because there are some intermingled things. In particular, the loop to count the total bytes also creates the ack handlers. While this avoids two loops, it also makes the process much more coupled and harder to reason through. Could we pull out the byte counting and flow control reservation into a separate method?